### DISCUSSION TITLE
Hard-replace macros

### DISCUSSION THREAD

**Discussion by andjar** [TiddlyWiki/TiddlyWiki5#5724] [Ideas]
Based on the [discussion on google groups](https://groups.google.com/g/tiddlywiki/c/crnjiJ4Spq4), there seems to be an interest in the feature of having a macro replaced with its output when a tiddler is saved. For example, if you want a timestamp, you could just write `<<<now>>>` and then the macro is replaced with the actual time when you save the tiddler (or already during edit if possible?). In general, `<<<macro>>>` would be replaced by the wikified output of `<<macro>>`. This is similar to [template substitution in mediawiki](https://en.wikipedia.org/wiki/Wikipedia:Substitution). This way, you are guaranteed that the content will not change if you send the tiddler to another wiki, edit the macro, or want a simple way fine-tune the macro output without editing the macro itself

**Comment by Jermolene** [comment]
I've been following those discussions with interest. I like the idea of being able to perform custom transformations on the text of a tiddler when saving it.

We could go further by also allowing for transformations that are applied when a text area is initialised with the text of a tiddler. That should make it possible to build an editor that edits a tiddler and its fields in a single textarea using ".tid" file format.

If possible, the transformations should be one-to-many, so that the transformations applied to a single text area can generate multiple tiddlers. That would make it possible to present a bunch of tiddler values in a single textarea for ease of editing, just like a data tiddler.

**Reply by saqimtiaz** [reply]
> We could go further by also allowing for transformations that are applied when a text area is initialised with the text of a tiddler.

This would allow use cases like being able to transform a TiddlyWiki format date string (as used in the created/modified fields) into a format appropriate for `<input type="date">`

It would also be interesting to allow actions to be specified to be invoked before saving a tiddler, without having to edit shadow tiddlers.
